pkg/users/users.go

```diff
@@ -80,113 +83,121 @@ func DeleteUser(jxClient versioned.Interface, ns string, userName string) error
+// If a match is found it is returned. If no match is found or an error occurs nil is returned together with
+// the potential error.
+func ResolveJXUser(user *gits.GitUser, providerKey string, jxClient versioned.Interface, namespace string) (*jenkinsv1.User, error) {
+	if user == nil {
+		return nil, errors.New("a git user needs to be specified")
+	}
 
-	id = naming.ToValidValue(id)
-	if id != "" {
+	// need to convert the git login to a valid Kubernetes label. This label is not unique and multiple users
+	// can match the same label, hence we need to match the returns Jenkins X users.
+	id := naming.ToValidValue(user.Login)
 
+	// First try to find by label - this is much faster as it uses an index
+	if id != "" {
 		labelSelector := fmt.Sprintf("%s=%s", providerKey, id)
-
-		// First try to find by label - this is much faster as it uses an index
-		users, err := jxClient.JenkinsV1().Users(namespace).List(metav1.ListOptions{
-			LabelSelector: labelSelector,
-		})
+		jxUsers, err := jxClient.JenkinsV1().Users(namespace).List(metav1.ListOptions{LabelSelector: labelSelector})
 		if err != nil {
 			return nil, err
 		}
-		if len(users.Items) > 1 {
-			return nil, fmt.Errorf("more than one user found in users.jenkins.io with label %s, found %v", labelSelector,
-				users.Items)
-		} else if len(users.Items) == 1 {
-			return &users.Items[0], nil
+		matchedUser, err := matchGitUser(user, jxUsers, providerKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// return match found within labeled Jenkins X users
+		if matchedUser != nil {
+			return matchedUser, nil
 		}
 	}
 
 	// Next try without the label - this might occur if someone manually updated the list
-	users, err := jxClient.JenkinsV1().Users(namespace).List(metav1.ListOptions{})
+	jxUsers, err := jxClient.JenkinsV1().Users(namespace).List(metav1.ListOptions{})
 	if err != nil {
 		return nil, err
 	}
-	if id != "" {
 
-		possibles := make([]jenkinsv1.User, 0)
-		for _, u := range users.Items {
-			for _, a := range u.Spec.Accounts {
-				if a.Provider == providerKey && a.ID == id {
-					possibles = append(possibles, u)
-				}
-			}
-		}
-		if len(possibles) > 1 {
-			possibleUsers := make([]string, 0)
-			for _, p := range possibles {
-				possibleUsers = append(possibleUsers, p.Name)
-			}
-			return nil, fmt.Errorf("more than one user found in users.jenkins.io with login %s for provider %s, "+
-				"found %s", id, providerKey, possibleUsers)
-		} else if len(possibles) == 1 {
-			// Add the label for next time
-			found := &possibles[0]
-			if found.Labels == nil {
-				found.Labels = make(map[string]string)
-			}
-			found.Labels[providerKey] = id
-			found, err := jxClient.JenkinsV1().Users(namespace).Update(found)
-			if err != nil {
-				return nil, err
-			}
-			log.Logger().Infof("Adding label %s=%s to user %s in users.jenkins.io", providerKey, id, found.Name)
-			return found, nil
+	matchedUser, err := matchGitUser(user, jxUsers, providerKey)
```

Comment:
See above, this code is intentionally not directly connected to Git.

Reply:
one could return the list of users and let the caller do the matching. You also don't need to pass the full gits.GitUser, but just the id. But this would still require to adjust the Slack app.